integrate/simpsons.go

```diff
@@ -0,0 +1,133 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"sort"
+)
+
+// Simpson's method approximates the integral of a function f(x)
+// by means of subdividing the interval of integration into segments
+// and applying a method that fits a polynomial to each subinterval.
+// this implementation makes use of the following composite Simpson's method
+// to estimate \int_a^b f(x) dx:
+//
+// \sum_{i=0}^{(N/2)-1} {a_0}_i * f_{2i+1} + {a_1}_i * f_{2i+1} + {a_2}_i * f_{2i}
+//
+// where N is the number of subintervals and the number of subintervals
+// is even and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
+//
+// {a_0}_i = 2 * h^{3}_{2i+1} - h^{3}_{2i} + 3 * h_{2i} * h^{2}_{2i+1} /
+//           6 * h_{2i+1} * (h_{2i+1} + h_{2i})
+//
+// {a_1}_i = h^{3}_{2i+1} + h^{3}_{2i} +
+//           3 * h_{2i+1} * (h_{2i+1} + h_{2i}) /
+//           6 * h_{2i+1} * h_{2i}
+//
+// {a_2}_i = 2 * h^{3}_{2i} - h^{3}_{2i+1} + 3 * h_{2i+1} * h^{2}_{2i} /
+//           6 * h_{2i} * (h_{2i+1} + h_{2i})
+//
+// where h_{k} is the length of the subinterval at k.
+//
+// The formula above approximates the integral of function of f
+// if N is an even number of subintervals.
+//
+// If the number of subintervals are odd,
+// the subintervals up i=0..n-2 are given by the above
+// and the approximations over the second to last and last subintervals
+// are given by:
+//
+// {a_0}_i * f_{N} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N-2}
+//
+// where the coefficients are:
+//
+// {a_0}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
+//           6 * (h_{N-2} + h_{N-1})
+//
+// {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
+//           6 * h_{N-2}
+//
+// {a_2}_i = -1 * h^{3}_{N-1} /
+//            6 * h_{N-2} * (h_{N-2} + h_{N-1})
+// More information is available at:
+// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
+//
+// The (x,f) input data points must be sorted along x.
+// One can use github.com/gonum/stat.SortWeighted to do that.
+// The x and f slices must be of equal length and have length > 1.
+func Simpsons(x, f []float64) float64 {
+	validateInterval(x, f)
+
+	var integral float64
+
+	// If only three points are supplied,
+	// approximate the integral
+	// assuming the second coordinate points
+	// suffices as a midpoint and interpolate with Simpson's method.
+	if len(x) == 3 {
+
+		// Panic if either subinterval coordinate points match
+		if x[0] == x[1] && f[0] == f[1] || x[1] == x[2] && f[0] == f[1] {
```

Comment:
What does this check? Why this special case is more stringent than the general case? This does not compensate for the fact that the formula below ignores `x[1]`. I would remove this.

Reply:
The existing checks guarantee ordering and that three points are supplied, but if I pass the points (0,1), (0,1), (1,1) or (0,0), (1,1), (1,1) there's no middle point.

Reply:
Why is (0,1)(0,2,)(1,1) allowed? Why is this check for repeated locations done only in the 3-length case? Does it mean that `validateInterval` does not actually validate the input?

Reply:
I think we should just disallow repeated points altogether. It's easy to check as you loop through the slice. The algorithm doesn't really make sense with repeated x locations.

Reply:
> we should just disallow repeated points altogether

That's exactly where I'm trying to get with the questions. But then `sort.Float64sAreSorted` cannot be used which means adding a helper here and this PR is already taking too long. That's why I asked for removing the check here.

Reply:
Square waves and Haar wavelets and noisy data sets may produce conditions where x maps to multiple values?

Reply:
If you have multiple values for one x, then such a mapping is not a function. If your data points come from some noisy measurements and may contain repeated locations, then you must either clean the input first or approximate the data with a smooth function. All of this is out of scope for this package.